src/Mobs/Monster.cpp

```diff
@@ -252,6 +259,64 @@ void cMonster::MoveToWayPoint(cChunk & a_Chunk)
 
 
 
+bool cMonster::EnsureProperDestination(cChunk & a_Chunk)
+{
+	cChunk * Chunk = a_Chunk.GetNeighborChunk(m_FinalDestination.x, m_FinalDestination.z);
+	BLOCKTYPE BlockType;
+	NIBBLETYPE BlockMeta;
+	int RelX = m_FinalDestination.x - Chunk->GetPosX() * cChunkDef::Width;
+	int RelZ = m_FinalDestination.z - Chunk->GetPosZ() * cChunkDef::Width;
+	if ((Chunk != nullptr) && Chunk->IsValid())
+	{
+		// If destination in the air, go down to the lowest air block.
+		for (;;)
+		{
+			Chunk->GetBlockTypeMeta(RelX, m_FinalDestination.y - 1, RelZ, BlockType, BlockMeta);
+			if (cBlockInfo::IsSolid(BlockType))
+			{
+				break;
+			}
+			m_FinalDestination += Vector3d(0, -1, 0);
+		}
+
+
+		// If destination in water, go up to the highest water block.
+		// If destination in solid, go up to first air block.
+		bool InWater = false;
+		for (;;)
+		{
+			Chunk->GetBlockTypeMeta(RelX, m_FinalDestination.y, RelZ, BlockType, BlockMeta);
+			if (BlockType == E_BLOCK_STATIONARY_WATER)
+			{
+				InWater = true;
+			}
+			else if (cBlockInfo::IsSolid(BlockType))
+			{
+				InWater = false;
+			}
+			else
+			{
+				break;
+			}
+			m_FinalDestination += Vector3d(0, 1, 0);
+		}
```

Comment:
Both loops will fail in a world where the void below is accessible - if there is no solid block under the mob. Use `while (m_FinalDestination.y < cChunkDef::Height)`